src/index/sparse/sparse_inverted_index.h

```diff
@@ -0,0 +1,449 @@
+        for (auto [i, v] : q_vec) {
+            for (size_t j = 0; j < inverted_lut_[i].size(); j++) {
+                auto [idx, val] = inverted_lut_[i][j];
+                scores[idx] += v * float(val);
+            }
+        }
+        for (size_t i = 0; i < n_rows_; ++i) {
+            if ((bitset.empty() || !bitset.test(i)) && scores[i] != 0) {
+                heap.push(i, -scores[i]);
+            }
+        }
+    }
+
+    class Cursor {
+     public:
+        Cursor() = default;
+        Cursor(const std::vector<Neighbor<T>>* lut, size_t num_vec, float max_score, float q_value,
+               const BitsetView* bitset)
+            : lut_(lut), num_vec_(num_vec), max_score_(max_score), q_value_(q_value), bitset_(bitset) {
+            while (loc_ < lut_->size() && !bitset_->empty() && bitset_->test(cur_vec_id())) {
+                loc_++;
+            }
+        }
+        Cursor(const Cursor& rhs) = delete;
+        Cursor(Cursor&& rhs) noexcept {
+            swap(rhs, *this);
+        }
+        Cursor&
+        operator=(const Cursor& rhs) = delete;
+        Cursor&
+        operator=(Cursor&& rhs) noexcept {
+            Cursor tmp(std::move(rhs));
+            swap(*this, tmp);
+            return *this;
+        }
+        void
+        next() {
+            loc_++;
+            while (loc_ < lut_->size() && !bitset_->empty() && bitset_->test(cur_vec_id())) {
+                loc_++;
+            }
+        }
+        // advance loc until cur_vec_id() >= vec_id
+        void
+        seek(table_t vec_id) {
+            while (loc_ < lut_->size() && cur_vec_id() < vec_id) {
+                next();
+            }
+        }
+        [[nodiscard]] table_t
+        cur_vec_id() const {
+            if (is_end()) {
+                return num_vec_;
+            }
+            return (*lut_)[loc_].id;
+        }
+        T
+        cur_distance() const {
+            return (*lut_)[loc_].distance;
+        }
+        [[nodiscard]] bool
+        is_end() const {
+            return loc_ >= size();
+        }
+        [[nodiscard]] float
+        q_value() const {
+            return q_value_;
+        }
+        [[nodiscard]] size_t
+        size() const {
+            return lut_->size();
+        }
+        [[nodiscard]] float
+        max_score() const {
+            return max_score_;
+        }
+
+     private:
+        friend void
+        swap(Cursor& lhs, Cursor& rhs) {
+            using std::swap;
+            swap(lhs.lut_, rhs.lut_);
+            swap(lhs.loc_, rhs.loc_);
+            swap(lhs.num_vec_, rhs.num_vec_);
+            swap(lhs.max_score_, rhs.max_score_);
+            swap(lhs.q_value_, rhs.q_value_);
+            // all cursors share the same bitset
+        }
+        const std::vector<Neighbor<T>>* lut_;
+        size_t loc_ = 0;
+        size_t num_vec_ = 0;
+        float max_score_ = 0.0f;
+        float q_value_ = 0.0f;
+        const BitsetView* bitset_;
```

Comment:
BitsetView is a lightweight structure (a wrapper around the pointer, basically) that should be passed by reference